apps/s_client.c

```diff
@@ -2322,7 +2344,31 @@ int s_client_main(int argc, char **argv)
             BIO *fbio = BIO_new(BIO_f_buffer());
 
             BIO_push(fbio, sbio);
-            BIO_printf(fbio, "CONNECT %s HTTP/1.0\r\n\r\n", connectstr);
+            BIO_printf(fbio, "CONNECT %s HTTP/1.0\r\n", connectstr);
+	    /* 
+	     * Workaround for broken proxies which would otherwise close
+	     * the connection when entering tunnel mode (eg Squid 2.6)
+	     */
+	    BIO_printf(fbio, "Proxy-Connection: Keep-Alive\r\n");
+
+            /* Support for basic (base64) proxy authentication */
+	    if (proxyuser != NULL) {
+		size_t l;
+		char *proxyauth, *proxyauthenc;
+	
+		l = strlen(proxyuser);
+		if (proxypass != NULL)
+		    l += strlen(proxypass); 
+		proxyauth = app_malloc(l+2, "Proxy auth string");
+		snprintf(proxyauth, l+2, "%s:%s", proxyuser, (proxypass != NULL) ? proxypass:"");
+		proxyauthenc = base64encode(proxyauth, strlen(proxyauth));
+		BIO_printf(fbio, "Proxy-Authorization: Basic %s\r\n", proxyauthenc); 
+		OPENSSL_clear_free(proxyauth, strlen(proxyauth));
+		OPENSSL_clear_free(proxyauthenc, strlen(proxyauthenc));
```

Comment:
Perhaps the password should be cleared too?

Reply:
Good point. I've added this after the end: label to catch all cases.